File: samples/java/com/jungo/shared/DiagLib.java
package com.jungo.shared;

import com.jungo.wdapi.*;
import java.util.Scanner;

/* Jungo Connectivity Confidential. Copyright (c) 2022 Jungo Connectivity Ltd.  https://www.jungo.com */

public class DiagLib {

    final public static int
    DIAG_INVALID_MENU = -1,
    DIAG_INPUT_CANCEL = -1,
    DIAG_INPUT_FAIL = 0,
    DIAG_INPUT_SUCCESS = 1;

    final public static char DIAG_CANCEL = 'x';

    final public static int
    DIAG_EXIT_MENU = 99,
    BYTES_IN_LINE = 32,
    HEX_CHARS_PER_BYTE = 2,
    HEX_STOP_POS = BYTES_IN_LINE * HEX_CHARS_PER_BYTE;

    final public static Scanner scanner = new Scanner(System.in);

    final public static int MENU_MAX_CHILDREN = 16;

    public interface MenuCallback {
        public long run(Object pCbCtx);
    }

    public interface MenuIsHiddenCallback {
        public boolean run(DiagMenuOption pMenu);
    }

    public static class DiagMenuOption {
        public String cOptionName; /**< The name of the option that will be
                                printed before selecting this menu */
        public String cTitleName; /**< The title name that will be printed at
                                the top of the menu(can be empty) */

        public MenuCallback cbEntry; /**< Callback function that runs when entering
                                    this menu */
        public MenuCallback cbExit;  /**< Callback function that runs when exiting
                                    this menu */
        public Object pCbCtx;               /**< The context needed for this menu
                                        (varibles that the menu needs to use)*/

        public boolean fIsHidden;  /**< Determine if this option should be visible
                            to the user, depending on its context*/
        public MenuIsHiddenCallback cbIsHidden; /**< Callback that returns whether
										        this menu should be hidden,
										        depending on the context.
										        This function is called
										        each time inside the menu loop */


        public int dwCurrentNumberOfChildren; /**< Current number of child menus in
                                             the pChildMenus array */
        public DiagMenuOption[] pChildMenus; /**< Array of all
                                           menus that can be
                                           accessed from this
                                           menu */
        public DiagMenuOption pParentMenu; /**< Parent of this menu. When
                           exiting the menu, the parent
                           menu will be shown */

        public DiagMenuOption(String cOptionName, String cTitleName,
                MenuCallback cbEntry, MenuCallback cbExit,
                MenuIsHiddenCallback cbIsHidden)
        {
            this.cOptionName = cOptionName;
            this.cTitleName = cTitleName;
            this.cbEntry = cbEntry;
            this.cbExit = cbExit;
            this.cbIsHidden = cbIsHidden;
            this.dwCurrentNumberOfChildren = 0;
            this.pChildMenus = new DiagMenuOption[MENU_MAX_CHILDREN];
            this.pParentMenu = null;
            this.pCbCtx = null;
        }

        public void DIAG_AddMenuOptionChild(DiagMenuOption pChildMenu)
        {
            if (pChildMenu != null)
            {
                if (this.dwCurrentNumberOfChildren < MENU_MAX_CHILDREN)
                {
                    this.pChildMenus[this.dwCurrentNumberOfChildren] =
                        pChildMenu;
                    this.dwCurrentNumberOfChildren++;
                    pChildMenu.pParentMenu = this;
                }
                else
                {
                    System.err.printf("DIAG_AddMenuOptionChild: %s",
                            "Error! children menu array is full.");
                }
            }
        }
    }

    // Builder class for DiagMenuOption, to simplify object creation
    public static class DiagMenuOptionBuilder
    {
        String _cOptionName = "";
        String _cTitleName = "";
        MenuCallback _cbEntry = null;
        MenuCallback _cbExit = null;
        MenuIsHiddenCallback _cbIsHidden = null;

        public DiagMenuOption build()
        {
            return new DiagMenuOption(_cOptionName, _cTitleName, _cbEntry,
                    _cbExit, _cbIsHidden);
        }

        public DiagMenuOptionBuilder cOptionName(String _cOptionName)
        {
            this._cOptionName = _cOptionName;
            return this;
        }

        public DiagMenuOptionBuilder cTitleName(String _cTitleName)
        {
            this._cTitleName = _cTitleName;
            return this;
        }

        public DiagMenuOptionBuilder cbEntry(MenuCallback _cbEntry)
        {
            this._cbEntry = _cbEntry;
            return this;
        }

        public DiagMenuOptionBuilder cbExit(MenuCallback _cbExit)
        {
            this._cbExit = _cbExit;
            return this;
        }

        public DiagMenuOptionBuilder cbIsHidden(
                MenuIsHiddenCallback _cbIsHidden)
        {
            this._cbIsHidden = _cbIsHidden;
            return this;
        }
    }

    public static void DIAG_MenuSetCtxAndParentForMenus(
            DiagMenuOption[] pMenusArr, Object pCtx,
            DiagMenuOption pParentMenu)
    {
        for (DiagMenuOption pMenu : pMenusArr)
        {
            if (pMenu != null)
            {
                pMenu.pCbCtx = pCtx;
                pParentMenu.DIAG_AddMenuOptionChild(pMenu);
            }
        }
    }

    private static void DIAG_MenuSetHiddenChildren(DiagMenuOption pMenu)
    {
        if (pMenu != null)
        {
            for (DiagMenuOption pChild : pMenu.pChildMenus)
            {
                if (pChild != null && pChild.cbIsHidden != null)
                	pChild.fIsHidden = pChild.cbIsHidden.run(pChild);
            }
        }
    }

    private static int DIAG_GetNextMenuOption(DiagMenuOption pCurrentMenu)
    {
        int dwOption = 0, dwStatus = 0;
        int dwMaxOption = pCurrentMenu.dwCurrentNumberOfChildren;

        System.out.println("Enter option: ");

        try
        {
            if (scanner.hasNext())
                dwOption = scanner.nextInt();
        }
        catch (Exception e)
        {
            System.out.printf("Invalid option\n");
            scanner.next();
            return DIAG_INVALID_MENU;
        }

        if (0 < dwOption && dwOption <= dwMaxOption)
        {
            if (pCurrentMenu.pChildMenus[dwOption - 1].fIsHidden)
                dwOption = DIAG_INVALID_MENU;
            else
                dwOption = dwOption - 1;
        }

        else if (dwOption != DIAG_EXIT_MENU)
        {
            dwOption = DIAG_INVALID_MENU;
        }

        if (dwOption == DIAG_INVALID_MENU)
            System.out.println("Invalid option");

        return dwOption;
    }

    private static void DIAG_MenuPrintTitleSeperator(String pcTitleName)
    {
        for(int i = 0; i < pcTitleName.length(); i++)
                System.out.print("-");

        System.out.println();
    }

    private static void DIAG_MenuPrintOptions(DiagMenuOption pMenu)
    {
        DiagMenuOption[] pChildren;

        if (pMenu != null)
        {
            pChildren = pMenu.pChildMenus;

            for (int i = 0; i < MENU_MAX_CHILDREN && pChildren[i] != null; i++)
            {
                if (!pChildren[i].fIsHidden)
                {
                    System.out.printf("%d. %s\n", (i + 1),
                        pChildren[i].cOptionName);
                }
            }

            System.out.printf("%d. Exit Menu\n", DIAG_EXIT_MENU);
            System.out.printf("\n");
        }
    }

    public static long DIAG_MenuRun(DiagMenuOption pMenuRoot)
    {
        DiagMenuOption pCurrentMenu = pMenuRoot;
        long dwOption, dwStatus = 0;

        while (pCurrentMenu != null)
        {
            dwStatus = 0;

            if (pCurrentMenu.cbEntry != null)
               dwStatus = pCurrentMenu.cbEntry.run(pCurrentMenu.pCbCtx);

            /* If error occurred, or we are in leaf node, go back to parent */
            if (dwStatus != 0 || pCurrentMenu.dwCurrentNumberOfChildren == 0)
            {
                pCurrentMenu = pCurrentMenu.pParentMenu;
                continue;
            }

            if (!pCurrentMenu.cTitleName.isEmpty())
            {
                System.out.printf("\n%s\n", pCurrentMenu.cTitleName);
                DIAG_MenuPrintTitleSeperator(pCurrentMenu.cTitleName);
            }

            DIAG_MenuSetHiddenChildren(pCurrentMenu);
            DIAG_MenuPrintOptions(pCurrentMenu);

            dwOption = DIAG_GetNextMenuOption(pCurrentMenu);
            if (dwOption == DIAG_EXIT_MENU)
            {
                if (pCurrentMenu.cbExit != null)
                    dwStatus = pCurrentMenu.cbExit.run(pCurrentMenu.pCbCtx);

                /* Go back to parent */
                pCurrentMenu = pCurrentMenu.pParentMenu;
            }
            else if (dwOption != DIAG_INVALID_MENU)
            {
                /* Go to chosen child */
                pCurrentMenu = pCurrentMenu.pChildMenus[(int) dwOption];
            }
        }

        return dwStatus;
    }

    // Mimic C __FUNCTION__ macro, will return the calling(!) function name
    public static String __FUNCTION__()
    {
        return Thread.currentThread().getStackTrace()[2].getMethodName();
    }

    public static void DIAG_PrintHexBuffer(byte[] pbData)
    {
        String buf = bytesToHex(pbData);
        int dwOffset, dwLineOffset;

        if (buf.isEmpty())
        {
            System.err.printf("DIAG_PrintHexBuffer: Error - %s\n",
                "Empty buffer");
            return;
        }

        for (dwOffset = 0; dwOffset < buf.length(); dwOffset++)
        {
            dwLineOffset = dwOffset % BYTES_IN_LINE;
            if (dwOffset != 0 && dwLineOffset == 0)
            {
                System.out.println();
            }
            if (dwOffset % HEX_CHARS_PER_BYTE == 0 && dwLineOffset > 0)
                System.out.print(" ");
            System.out.print(buf.charAt(dwOffset));

        }
    }

    public static int DIAG_GetHexChar()
    {
        int ch;
        try {
            ch = System.in.read();
        }
        catch (Exception e)
        {
            return -1;
        }

        if (Character.digit((char)ch, 16) == -1)
            return -1;

        if (Character.isDigit(ch))
            return ch - '0';
        else
            return Character.toUpperCase(ch) - 'A' + 10;
    }

    public static byte[] DIAG_GetHexBuffer(int dwBytes)
    {
        int dwBytesRead = 0;
        int res;
        int ch;
        byte[] pData = new byte[dwBytes];

        while (dwBytesRead < dwBytes)
        {
            ch = DIAG_GetHexChar();
            if (ch < 0)
                continue;

            res = ch << 4;

            ch = DIAG_GetHexChar();
            if (ch < 0)
                continue;

            res += ch;
            pData[dwBytesRead] = (byte)res;
            dwBytesRead++;
        }

        /* Advance to new line */
        do {
            try {
                ch = System.in.read();
            }
            catch (Exception e)
            {
                return pData;
            }
        } while (ch != '\n' && ch != '\r');

        /* Return the number of bytes that was read */
        return pData;
    }

    /* Get menu option from user */
    public static WDCResultInteger DIAG_GetMenuOption(int iMax)
    {
        int iRet = 0;

        System.out.printf("Enter option: ");

        try
        {
            if (scanner.hasNext())
                iRet = scanner.nextInt();
        }
        catch (Exception e)
        {
            System.out.printf("Invalid option\n");
            scanner.next();
            return new WDCResultInteger(DIAG_INPUT_FAIL, 0);
        }
        if (iRet < 1)
        {
            System.out.printf("Invalid option\n");
            return new WDCResultInteger(DIAG_INPUT_FAIL, 0);
        }

        if (iRet == DIAG_EXIT_MENU)
            return new WDCResultInteger(DIAG_INPUT_SUCCESS, DIAG_EXIT_MENU);

        if (iRet > iMax)
        {
            System.out.printf("Invalid option: Option must be %s%d, or %d to" +
                " exit\n", (1 == iMax) ? "" : "between 1 - ", iMax,
                DIAG_EXIT_MENU);
            return new WDCResultInteger(DIAG_INPUT_FAIL, 0);
        }

        return new WDCResultInteger(DIAG_INPUT_SUCCESS, iRet);
    }

    public static WDCResultLong DIAG_InputNum(String sInputText, boolean fHex,
        long min, long max)
    {
        long dwRet = 0;
        boolean fCheckRange = (max > min);
        String s;

        System.out.printf("%s (to cancel press 'x'): %s",
            (sInputText == null || sInputText.isEmpty()) ? "Enter input" :
            sInputText, fHex ? "0x" : "");
        try
        {
            s = scanner.next();
            dwRet = Long.parseUnsignedLong(s, fHex ? 16 : 10);
        }
        catch (Exception e)
        {
            System.out.printf("Input cancelled / invalid\n");
            return new WDCResultLong(DIAG_INPUT_FAIL, -1);
        }

        if (fCheckRange)
        {
            if (dwRet < min || dwRet > max)
            {
                System.out.printf("Invalid input: Input must be between ");
                if (fHex)
                    System.out.printf("0x%x and 0x%x\n", min, max);
                else
                    System.out.printf("%d and %d\n", min, max);
                return new WDCResultLong(DIAG_INPUT_FAIL, -1);
            }
        }

        return new WDCResultLong(DIAG_INPUT_SUCCESS, dwRet);
    }

    private final static char[] hexArray = "0123456789ABCDEF".toCharArray();
    public static String bytesToHex(byte[] bytes)
    {
        char[] hexChars = new char[bytes.length * 2];
        for ( int j = 0; j < bytes.length; j++ ) {

            int v = bytes[j] & 0xFF;
            hexChars[j * 2] = hexArray[v >>> 4];
            hexChars[j * 2 + 1] = hexArray[v & 0x0F];
        }
        return new String(hexChars) + "\n";
    }

    /*
     * Function: get_cur_time() Retrieves the current time, in OS units
     * Parameters: time [out] pointer to the OS time, in OS units Return Value:
     * true if successful, false otherwise
     */
    public static long get_cur_time()
    {
        return System.currentTimeMillis();
    }

    /*
     * Function: time_diff() Calculates a time difference in milliseconds
     * Parameters: end [in] End time, in OS units start [in] Start time, in OS
     * units Return Value: The time difference, in milliseconds.
     */
    private static long time_diff(final long end, final long start)
    {
        return end - start;
    }

    public static void DIAG_PrintPerformance(long size, long startTime)
    {
        long endTime = get_cur_time();
        long perf_time_total = time_diff(endTime, startTime);
        if (perf_time_total != -1)
        {
            System.out.printf("Transferred %d bytes, elapsed time %d[ms], " +
                "rate %d[MB/sec]\n", size, perf_time_total,
                /* (bytes / msec) * sec / MB */
                (size / (perf_time_total + 1)) * 1000 / (1024 * 1024));
        }
    }
}

File: samples/c/shared/diag_lib.c
/* Jungo Connectivity Confidential. Copyright (c) 2022 Jungo Connectivity Ltd.  https://www.jungo.com */

/******************************************************************************
*  File: diag_lib.c - Implementation of shared WD user-mode diagnostics API.
*
*  Note: This code sample is provided AS-IS and as a guiding sample only.
*******************************************************************************/

#if !defined(__KERNEL__)

#include <stdio.h>
#include <time.h>
#include "diag_lib.h"

/*************************************************************
  General definitions
 *************************************************************/
/* Error messages display */
#define DIAG_ERR printf

/* Messages display */
#define DIAG_TRACE printf

/*************************************************************
  Functions implementation
 *************************************************************/
static void DIAG_AddMenuOptionChild(
    DIAG_MENU_OPTION *pParentMenu, DIAG_MENU_OPTION *pChildMenu)
{
    if (pParentMenu && pChildMenu)
    {
        if (pParentMenu->dwCurrentNumberOfChildren < MENU_MAX_CHILDREN)
        {
            pParentMenu->pChildMenus[pParentMenu->dwCurrentNumberOfChildren] =
                pChildMenu;
            pParentMenu->dwCurrentNumberOfChildren++;
            pChildMenu->pParentMenu = pParentMenu;
        }
        else
        {
            DIAG_ERR("DIAG_AddMenuOptionChild: Error! children menu array is "
                "full.");
        }
    }
}

void DIAG_MenuSetCtxAndParentForMenus(DIAG_MENU_OPTION *pMenusArr,
    DWORD dwSize, PVOID pCtx, DIAG_MENU_OPTION *pParentMenu)
{
    DWORD i;

    for (i = 0; i < dwSize; i++)
    {
        pMenusArr[i].pCbCtx = pCtx;
        DIAG_AddMenuOptionChild(pParentMenu, &pMenusArr[i]);
    }
}

static void DIAG_MenuSetHiddenChildren(DIAG_MENU_OPTION *pMenu)
{
    DWORD i;
    DIAG_MENU_OPTION **ppChildren;

    if (pMenu)
    {
        ppChildren = pMenu->pChildMenus;

        for (i = 0; i < MENU_MAX_CHILDREN && ppChildren[i]; i++)
        {
            if (ppChildren[i]->cbIsHidden)
            {
                ppChildren[i]->fIsHidden =
                    ppChildren[i]->cbIsHidden(ppChildren[i]);
            }
        }
    }
}

static DWORD DIAG_GetNextMenuOption(DIAG_MENU_OPTION *pCurrentMenu)
{
    static CHAR sInput[MAX_NAME];
    DWORD dwStatus, dwOption, dwMaxOption;

    dwMaxOption = pCurrentMenu->dwCurrentNumberOfChildren;
    dwOption = 0;

    DIAG_TRACE("Enter option: ");

    fgets(sInput, sizeof(sInput), stdin);

    dwStatus = sscanf(sInput, "%u", &dwOption);
    if (dwStatus < 1 || dwOption == 0)
        goto Error;

    if (0 < dwOption && dwOption <= dwMaxOption)
    {
        if (pCurrentMenu->pChildMenus[dwOption - 1]->fIsHidden)
            goto Error;

        dwOption = dwOption - 1;
    }
    else if (dwOption != DIAG_EXIT_MENU)
    {
        goto Error;
    }

    goto Exit;

Error:
    DIAG_TRACE("Invalid option\n");
    dwOption = DIAG_INVALID_MENU;

Exit:
    return dwOption;
}

static void DIAG_MenuPrintTitleSeperator(PCHAR pcTitleName)
{
    DWORD i = 0;

    while (pcTitleName[i++])
        DIAG_TRACE("-");

    DIAG_TRACE("\n");
}

static void DIAG_MenuPrintOptions(DIAG_MENU_OPTION *pMenu)
{
    DWORD i;
    DIAG_MENU_OPTION **ppChildren;

    if (pMenu)
    {
        ppChildren = pMenu->pChildMenus;

        for (i = 0; i < MENU_MAX_CHILDREN && ppChildren[i]; i++)
        {
            if (!ppChildren[i]->fIsHidden)
                DIAG_TRACE("%d. %s\n", (i + 1), ppChildren[i]->cOptionName);
        }

        DIAG_TRACE("%d. Exit Menu\n", DIAG_EXIT_MENU);
        DIAG_TRACE("\n");
    }
}

DWORD DIAG_MenuRun(DIAG_MENU_OPTION *pMenuRoot)
{
    DIAG_MENU_OPTION *pCurrentMenu = pMenuRoot;
    DWORD dwOption, dwStatus = 0;

    while (pCurrentMenu)
    {
        dwStatus = 0;

        if (pCurrentMenu->cbEntry)
           dwStatus = pCurrentMenu->cbEntry(pCurrentMenu->pCbCtx);

        /* If error occured, or we are in leaf node, go back to parent */
        if (dwStatus || pCurrentMenu->dwCurrentNumberOfChildren == 0)
        {
            pCurrentMenu = pCurrentMenu->pParentMenu;
            continue;
        }

        if (*pCurrentMenu->cTitleName)
        {
            DIAG_TRACE("\n%s\n", pCurrentMenu->cTitleName);
            DIAG_MenuPrintTitleSeperator(pCurrentMenu->cTitleName);
        }


        DIAG_MenuSetHiddenChildren(pCurrentMenu);
        DIAG_MenuPrintOptions(pCurrentMenu);

        dwOption = DIAG_GetNextMenuOption(pCurrentMenu);
        if (dwOption == DIAG_EXIT_MENU)
        {
            if (pCurrentMenu->cbExit)
                dwStatus = pCurrentMenu->cbExit(pCurrentMenu->pCbCtx);

            /* Go back to parent */
            pCurrentMenu = pCurrentMenu->pParentMenu;
        }
        else if (dwOption != DIAG_INVALID_MENU)
        {
            /* Go to chosen child */
            pCurrentMenu = pCurrentMenu->pChildMenus[dwOption];
        }
    }

    return dwStatus;
}

#define BYTES_IN_LINE 16
#define HEX_CHARS_PER_BYTE 3
#define HEX_STOP_POS BYTES_IN_LINE * HEX_CHARS_PER_BYTE
void DIAG_PrintHexBuffer(PVOID pBuf, DWORD dwBytes, BOOL fAscii)
{
    PBYTE pbData = (PBYTE)pBuf;
    CHAR pHex[HEX_STOP_POS + 1] = { 0 };
    CHAR pAscii[BYTES_IN_LINE + 1] = { 0 };
    DWORD dwOffset, dwLineOffset, i;

    if (!pBuf || !dwBytes)
    {
        DIAG_ERR("DIAG_PrintHexBuffer: Error - %s\n",
            !pBuf ? "NULL buffer pointer" : "Empty buffer");
        return;
    }

    for (dwOffset = 0; dwOffset < dwBytes; dwOffset++)
    {
        dwLineOffset = dwOffset % BYTES_IN_LINE;
        if (dwOffset && !dwLineOffset)
        {
            pAscii[dwLineOffset] = '\0';
            printf("%s%s%s\n", pHex, fAscii ? " | " : "", fAscii ? pAscii : "");
        }
        sprintf(pHex + dwLineOffset * HEX_CHARS_PER_BYTE, "%02X ",
            (UINT32)pbData[dwOffset]);
        if (fAscii)
        {
            pAscii[dwLineOffset] =
                (CHAR)((pbData[dwOffset] >= 0x20) ? pbData[dwOffset] : '.');
        }
    }

    /* Print the last line. Fill with blanks if needed */
    if (dwOffset % BYTES_IN_LINE)
    {
        for (i = (dwOffset % BYTES_IN_LINE) * HEX_CHARS_PER_BYTE;
            i < BYTES_IN_LINE * HEX_CHARS_PER_BYTE;
            i++)
        {
            pHex[i] = ' ';
        }
        pHex[i] = '\0';
    }
    if (fAscii)
        pAscii[dwOffset % BYTES_IN_LINE] = '\0';
    printf("%s%s%s\n", pHex, fAscii ? " | " : "", fAscii ? pAscii : "");
}

int DIAG_GetHexChar(void)
{
    int ch;

    ch = getchar();

    if (!isxdigit(ch))
        return -1;

    if (isdigit(ch))
        return ch - '0';
    else
        return toupper(ch) - 'A' + 10;
}

DWORD DIAG_GetHexBuffer(PVOID pBuffer, DWORD dwBytes)
{
    DWORD dwBytesRead = 0;
    PBYTE pData = (PBYTE)pBuffer;
    int res;
    int ch;

    while (dwBytesRead < dwBytes)
    {
        ch = DIAG_GetHexChar();
        if (ch < 0)
            continue;

        res = ch << 4;

        ch = DIAG_GetHexChar();
        if (ch < 0)
            continue;

        res += ch;
        pData[dwBytesRead] = (BYTE)res;
        dwBytesRead++;
    }

    /* Advance to new line */
    do {
        ch = getchar();
    } while (ch != '\n' && ch != '\r');

    /* Return the number of bytes that was read */
    return dwBytesRead;
}

void DIAG_FillHexBuffer(PVOID pBuffer, DWORD dwBytes, BYTE ch, BOOL fRandNum)
{
    if (fRandNum)
    {
        srand((unsigned int)time(NULL));
        ch = (rand() % 255);
    }

    memset(pBuffer, ch, dwBytes);
}

/* Get menu option from user */
DIAG_INPUT_RESULT DIAG_GetMenuOption(PDWORD pdwOption, DWORD dwMax)
{
    static CHAR sInput[256];
    int iRet;

    if (!pdwOption)
    {
        DIAG_ERR("DIAG_GetMenuOption: Error - NULL option pointer\n");
        return DIAG_INPUT_FAIL;
    }

    printf("Enter option: ");

    fgets(sInput, sizeof(sInput), stdin);

    iRet = sscanf(sInput, "%u", pdwOption);
    if (iRet < 1)
    {
        printf("Invalid option\n");
        return DIAG_INPUT_FAIL;
    }

    if (DIAG_EXIT_MENU == *pdwOption)
        return DIAG_INPUT_SUCCESS;

    if (!dwMax)
        return DIAG_INPUT_SUCCESS;

    if (*pdwOption > dwMax)
    {
        printf("Invalid option: Option must be %s%d, or %d to exit\n",
            (1 == dwMax) ? "" : "between 1 - ", dwMax, DIAG_EXIT_MENU);
        return DIAG_INPUT_FAIL;
    }

    return DIAG_INPUT_SUCCESS;
}

DIAG_INPUT_RESULT DIAG_InputNum(PVOID pInput, const CHAR *sInputText, BOOL fHex,
    DWORD dwSize, UINT64 min, UINT64 max)
{
    static CHAR sInput[256];
    const CHAR *sFormat;
    int iRet;
    BOOL fCheckRange = (max > min);

    if (!pInput)
    {
        DIAG_ERR("DIAG_InputData: Error - NULL input pointer\n");
        return DIAG_INPUT_FAIL;
    }

    printf("%s (to cancel press '%c'): %s",
        (!sInputText || !strcmp(sInputText, "")) ?  "Enter input" : sInputText,
        DIAG_CANCEL, fHex ? "0x" : "");

    fgets(sInput, sizeof(sInput), stdin);
    if (DIAG_CANCEL == tolower(sInput[0]))
        return DIAG_INPUT_CANCEL;

    switch (dwSize)
    {
    case sizeof(WORD):
        sFormat = fHex ? "%hX" : "%hu";
        break;
    case sizeof(UINT32):
        sFormat = fHex ? "%X" : "%u";
        break;
    case sizeof(UINT64):
        sFormat = fHex ? "%" PRI64 "X" : "%" PRI64 "d";
        break;
    default:
        DIAG_ERR("DIAG_InputNum: Error - Invalid size (%d)\n", dwSize);
        return DIAG_INPUT_FAIL;
    }

    iRet = sscanf(sInput, sFormat, pInput);
    if (iRet < 1)
    {
        printf("Invalid input\n");
        return DIAG_INPUT_FAIL;
    }

    if (fCheckRange)
    {
        UINT64 tmp = sizeof(WORD) == dwSize ? *(WORD *)pInput :
            sizeof(UINT32) == dwSize ? *(UINT32 *)pInput : *(UINT64 *)pInput;

        if (tmp < min || tmp > max)
        {
            printf("Invalid input: Input must be between ");
            if (fHex)
                printf("0x%" PRI64 "X and 0x%" PRI64 "X\n", min, max);
            else
                printf("%" PRI64 "d and %" PRI64 "d\n", min, max);
            return DIAG_INPUT_FAIL;
        }
    }

    return DIAG_INPUT_SUCCESS;
}

/* Function: time_diff()
Calculates a time difference in milliseconds
Parameters:
end   [in] End time, in OS units
start [in] Start time, in OS units
Return Value:
The time difference, in milliseconds. */
double time_diff(TIME_TYPE *end, TIME_TYPE *start)
{
#if defined(UNIX)
    return (double)(end->tv_usec - start->tv_usec) / 1000 +
        (end->tv_sec - start->tv_sec) * 1000;
#else
    TIME_TYPE ctr_freq;

    if (!QueryPerformanceFrequency(&ctr_freq))
    {
        DIAG_ERR("Error reading timer frequency\n");
        return -1;
    }

    return (double)((end->QuadPart - start->QuadPart) * 1000 /
        ctr_freq.QuadPart);
#endif
}

/* Function: get_cur_time()
Retrieves the current time, in OS units
Parameters:
time [out] pointer to the OS time, in OS units
Return Value:
TRUE if successful, FALSE otherwise */
BOOL get_cur_time(TIME_TYPE *time)
{
#if defined(UNIX)
    return !gettimeofday(time, NULL);
#else
    return QueryPerformanceCounter(time);
#endif
}

void DIAG_PrintPerformance(UINT64 qwBytes, TIME_TYPE *startTime)
{
    TIME_TYPE endTime;
    double perf_time_total;

    get_cur_time(&endTime);
    perf_time_total = time_diff(&endTime, startTime);

    if (perf_time_total != -1)
    {
        printf("Transferred %" PRI64 "d bytes, elapsed time %.2lf[ms], "
            "rate %.2lf [MB/sec]\n", qwBytes, perf_time_total,
            ((qwBytes / (perf_time_total + 1)) * 1000) / (1024 * 1024));
            /* ((bytes / msec) * sec) / MB */
    }
}

#endif /* !defined(__KERNEL__) */

